File: forge/src/main/java/juuxel/adorn/platform/forge/FluidBridgeForge.java
package juuxel.adorn.platform.forge;

import juuxel.adorn.fluid.FluidAmountPredicate;
import juuxel.adorn.fluid.FluidUnit;
import juuxel.adorn.fluid.FluidVolume;
import juuxel.adorn.platform.FluidBridge;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.core.component.DataComponentPatch;
import net.minecraft.world.level.material.Fluid;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.world.level.Level;
import net.neoforged.neoforge.capabilities.Capabilities;
import net.neoforged.neoforge.transfer.fluid.FluidResource;
import net.neoforged.neoforge.transfer.transaction.Transaction;
import org.jetbrains.annotations.Nullable;

public final class FluidBridgeForge implements FluidBridge {
    @Override
    public FluidUnit getFluidUnit() {
        return FluidUnit.LITRE;
    }

    @Nullable
    @Override
    public FluidVolume drain(Level world, BlockPos pos, @Nullable BlockState state, Direction side, Fluid fluid, FluidAmountPredicate amountPredicate) {
        // This method is a port of the Fabric fluid bridge code.
        var fluidHandler = world.getCapability(Capabilities.Fluid.BLOCK, pos, state, null, side);

        if (fluidHandler != null) {
            var upperBound = amountPredicate.getUpperBound();
            int maxAmount = (int) FluidUnit.convert(upperBound.getAmount(), upperBound.getUnit(), FluidUnit.LITRE);

            try (var tx = Transaction.open(null)) {
                var extracted = fluidHandler.extract(FluidResource.of(fluid), maxAmount, tx);

                if (extracted > 0 && amountPredicate.test(extracted, FluidUnit.LITRE)) {
                    tx.commit();
                    return new FluidVolume(fluid, extracted, DataComponentPatch.EMPTY, FluidUnit.LITRE);
                }
            }
        }

        return null;
    }
}

File: forge/src/main/kotlin/juuxel/adorn/platform/forge/FluidBridgeForge.kt
package juuxel.adorn.platform.forge

import juuxel.adorn.fluid.FluidAmountPredicate
import juuxel.adorn.fluid.FluidUnit
import juuxel.adorn.fluid.FluidVolume
import juuxel.adorn.platform.FluidBridge
import juuxel.adorn.platform.forge.util.toFluidVolume
import net.minecraft.block.BlockState
import net.minecraft.fluid.Fluid
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.world.World
import net.neoforged.neoforge.capabilities.Capabilities
import net.neoforged.neoforge.fluids.FluidStack
import net.neoforged.neoforge.fluids.capability.IFluidHandler

class FluidBridgeForge : FluidBridge {
    override val fluidUnit = FluidUnit.LITRE

    override fun drain(world: World, pos: BlockPos, state: BlockState?, side: Direction, fluid: Fluid, amountPredicate: FluidAmountPredicate): FluidVolume? {
        // This method is a port of the Fabric fluid bridge code.
        val fluidHandler = world.getCapability(Capabilities.FluidHandler.BLOCK, pos, state, null, side)

        if (fluidHandler != null) {
            val upperBound = amountPredicate.upperBound
            val maxAmount = FluidUnit.convert(upperBound.amount, from = upperBound.unit, to = FluidUnit.LITRE).toInt()
            val max = FluidStack(fluid, maxAmount)
            val extracted = fluidHandler.drain(max, IFluidHandler.FluidAction.SIMULATE)

            if (!extracted.isEmpty && amountPredicate.test(extracted.amount.toLong(), FluidUnit.LITRE)) {
                fluidHandler.drain(extracted, IFluidHandler.FluidAction.EXECUTE)
                return extracted.toFluidVolume()
            }
        }

        return null
    }
}
